Add Skew map

knelli2 opened the pull request:
## Proposed changes

Part of #4603.

![image](https://github.com/user-attachments/assets/4c63fb76-41c6-4c6d-9bd7-a20a757dcbcd)


<!--
At a high level, describe what this PR does.
-->

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

knelli2 committed:
Add Skew map

nilsvu commented on src/Domain/CoordinateMaps/TimeDependent/Skew.hpp:
```diff
@@ -0,0 +1,292 @@
+ * between the \p outer_radius and the ray connecting $\vec{x}_0$ and $\vec{x}$.
+ * Therefore, $|\vec{x}_P| = R$.
+ *
+ * \image html Skew.svg "Vectors for Skew map"
```
This file is relatively large for what it does with 111KB (our limit is 200KB). It's the handwriting as vector graphic that makes it large I think. I expect it would be much smaller as a PNG, could you try that?

nilsvu commented on src/Domain/CoordinateMaps/TimeDependent/Skew.hpp:
```diff
@@ -0,0 +1,292 @@
+ * system.
+ *
+ * ### Mapped Coordinates
+ * The Skew coordinte map is given by the mapping
```
typo "coordinate"

nilsvu commented on src/Domain/CoordinateMaps/TimeDependent/Skew.hpp:
```diff
@@ -0,0 +1,292 @@
+ * \end{equation}
+ *
+ * We can bound the root by noticing that in $\ref{eq:map}$, if we substitute
+ * the extramal values of $W = 0$ and $W = 1$, we get bounds on $\bar{x}$ which
```
typo "extremal"

nilsvu commented on src/Domain/CoordinateMaps/TimeDependent/Skew.hpp:
```diff
@@ -0,0 +1,292 @@
+ * \end{align}
+ *
+ * where $F_y(t)$ and $F_z(t)$ are the angles within the $(x,y)$ plane between
+ * the undistored $x$-axis and the skewed $\bar{y}$-axis at the \p center,
```
typo "undistorted"

nilsvu commented on src/Domain/CoordinateMaps/TimeDependent/Skew.hpp:
```diff
@@ -0,0 +1,292 @@
+ public:
+  static constexpr size_t dim = 3;
+
+  Skew(std::string function_of_time_name, const std::array<double, 3>& center,
```
Please add to the docs what expectations you have for the FoT. This info is also missing from other maps, but it's important because otherwise I don't know what numbers to put into the FoT to control the map.

nilsvu commented on src/Domain/CoordinateMaps/TimeDependent/Skew.cpp:
```diff
@@ -0,0 +1,423 @@
+template <typename T>
+std::array<tt::remove_cvref_wrap_t<T>, 3> Skew::operator()(
+    const std::array<T, 3>& source_coords, const double time,
+    const std::unordered_map<
```
(optional) `domain::FunctionsOfTimeMap` here and below

nilsvu commented on src/Domain/CoordinateMaps/TimeDependent/Skew.cpp:
```diff
@@ -0,0 +1,423 @@
+
+  // Short ciruit if we're at the outer radius
+  const double target_radius = magnitude(target_coords);
+  if (target_radius >= outer_radius_ and
```
why this first condition? or do you mean `or`?

knelli2 replied:
No, I'm checking if the radius is within roundoff of the outer radius, but strictly *outside* the outer radius. If it's epsilon inside the outer radius everything is fine and the point can just go through the map/inverse.

nilsvu replied:
If it's epsilon within the outer radius you can also short-circuit, no need to restrict the short-circuit to just one direction of the epsilon.

nilsvu commented on src/Domain/CoordinateMaps/TimeDependent/Skew.cpp:
```diff
@@ -0,0 +1,423 @@
+  ResultT& result = lambda;
+
+  for (size_t i = 0; i < get_size(result); i++) {
+    if (get_element(lambda, i) <= 0.0 and
```
Also here did you mean `or`?

nilsvu commented on src/Domain/CoordinateMaps/TimeDependent/Skew.cpp:
```diff
@@ -0,0 +1,423 @@
+    if (get_element(lambda, i) <= 0.0 and
+        equal_within_roundoff(get_element(lambda, i), 0.0)) {
+      get_element(result, i) = 1.0;
+    } else if (get_element(lambda, i) >= 1.0 and
```
Also here

nilsvu commented on src/Domain/CoordinateMaps/TimeDependent/Skew.cpp:
```diff
@@ -0,0 +1,423 @@
+      center_[1] * centered_source_coords_dot_center,
+      center_[2] * centered_source_coords_dot_center};
+  // We don't multiply grad_delta by 2 like in the dox because it will just
+  // canel with the factor of 1/2 in grad_width
```
typo "cancel"

nilsvu commented on src/Domain/CoordinateMaps/TimeDependent/Skew.cpp:
```diff
@@ -0,0 +1,423 @@
+      outer_radius_squared_minus_center_radius_squared_;
+  for (size_t i = 0; i < get_size(lambda); i++) {
+    // sin(lambda * M_PI) is zero at both lambda=0 and lambda=1
+    if ((get_element(lambda, i) <= 0.0 and
```
Also here `or` and below?

nilsvu commented on src/Domain/CoordinateMaps/TimeDependent/Skew.cpp:
```diff
@@ -0,0 +1,423 @@
+std::array<tt::remove_cvref_wrap_t<T>, 3> Skew::get_width_deriv(
+    const std::array<T, 3>& source_coords_cvref) const {
+  using ResultT = tt::remove_cvref_wrap_t<T>;
+  std::array<ResultT, 3> source_coords{};
```
Why do you need this non-const variable? Seems dangerous. Please add a comment if this is really necessary.

nilsvu commented on tests/Unit/Domain/CoordinateMaps/TimeDependent/Test_Skew.cpp:
```diff
@@ -0,0 +1,123 @@
+                                       f_of_t_list);
+    test_inverse_map(skew_map, point_xi, t, f_of_t_list);
+    test_inverse_map(skew_map_deserialized, point_xi, t, f_of_t_list);
+    run_checks(dv_point_xi);
```
Maybe check that the map gives expected results at some points that you know, like at the center and at the outer radius

knelli2 commented:
@nilsvu Pushed fixups and compressed the image some more. Wanted to keep it an svg so it rendered better online.

nilsvu commented on src/Domain/CoordinateMaps/TimeDependent/Skew.cpp:
```diff
@@ -232,18 +229,22 @@ template <typename T>
-      gsl::at(source_coords, i) = gsl::at(source_coords_cvref, i);
+  // Can't do math with the cvref typed data so we convert it to the result
+  // type. If it's a DataVector then we use views. If it's doubles, just a copy
+  const std::array<ResultT, 3> source_coords = [&source_coords_cvref]() {
```
Can you just do this:

```cpp
const auto& x = dereference_wrapper(source_coords_cvref[0]);
const auto& y = dereference_wrapper(source_coords_cvref[1]);
const auto& z = dereference_wrapper(source_coords_cvref[2]);
```

Then you don't need any of these const-casts etc.

nilsvu replied:
Edit: I think in other maps we just do math with `cref(DataVector)` and that seems to work, see Wedge.cpp. Haven't checked the details.

knelli2 replied:
No I want to do math with the entire `std::array`, not the individual components. However, that math won't work between a `std::array<DataVector>` and `std::array<std::reference_wrapper<const DataVector>>`. That's why I const cast.

nilsvu replied:
ok const-casts are pretty dangerous so maybe sprinkle some dereferencing in `StdArrayHelpers.hpp` instead?

knelli2 replied:
We use `const_cast` in a lot of places for this exact purpose (to set data references). Quick VSCode/git search shows 136 occurrences in 64 files

nilsvu commented on tests/Unit/Domain/CoordinateMaps/TimeDependent/Test_Skew.cpp:
```diff
@@ -113,11 +115,74 @@ void test(const gsl::not_null<Generator*> generator) {
+        skew_map.inverse(mapped_point, time, f_of_t_list);
+    CHECK(inverse_point.has_value());
+    CHECK_ITERABLE_APPROX(inverse_point.value(), test_point);
+    // Jacobian not defined at the center
```
do we not need it?

knelli2 replied:
Unclear. SpEC didn't handle this specially (and thus would get a division by zero error if the center was passed to the jacobian).

knelli2 commented:
Ok @nilsvu I updated this to address the concerns we had about undefined points and smoothness. I squashed all previous fixups together into one to make it easier to see what I changed

nilsvu commented on src/Domain/CoordinateMaps/TimeDependent/Skew.hpp:
```diff
@@ -59,76 +61,22 @@ namespace domain::CoordinateMaps::TimeDependent {
  * \begin{equation}
  *   \label{eq:lambda}
- *   \lambda(\vec{x}) = \frac{|\vec{x} - \vec{x}_0|}{|\vec{x}_P - \vec{x}_0|}
+ *   \lambda(\vec{x}) = \frac{|\vec{x}|^2}{R^2}.
```
My understanding is that you added the square to make this smooth - if that is the case, I don't see the reasoning there. The square only changes the degree of differentiability by 1, doesn't it? (Edit: this was wrong. $|\vec{x}|^2$ is smooth even though $|\vec{x}|$ is not.) The issue remains that $\frac{dr}{dx^i} = \frac{x^i}{r}$ is not smooth at the origin (e.g. along the x-axis it is -1 when approached from the left and +1 when approached from the right). I think this is a problem, unless we can guarantee that the origin is always on an element boundary.

knelli2 replied:
$W$ ***IS*** smooth though so it's fine. Also I've checked first 3 derivatives of $W$ are smooth as well (it's just sine terms multiplied by powers of $x$). Whether $\lambda$ is smooth is irrelevant. Only the derivatives of $W$ are the ones that mater.

nilsvu replied:
Ok I'm convinced now that the map is smooth. The square is actually not relevant because the cos is symmetric, so $\lambda(\vec{x}) = \frac{|\vec{x}|}{R}$ would also work.

markscheel replied:
I don't think `|x|/R` works, despite the cosine being symmetric.  Because when you differentiate, you get sine terms, and even worse, you get derivs of `|x|/R`).  `|x|^2/R^2` is fine though.

knelli2 commented:
@nilsvu I incorporated the change found in https://github.com/sxs-collaboration/spec/pull/2830 of subtracting off the center of the skew map in certain places, which was discussed with @markscheel. Now this should be smooth and correct at all points.

nilsdeppe commented on src/Domain/CoordinateMaps/TimeDependent/Skew.hpp:
```diff
@@ -0,0 +1,246 @@
+ * unchanged, but will distort (or skew) the $x$ coordinate based on functions
+ * of time and radial distance to the origin.
+ *
+ * \details This coordinate map is only available in 3 dimensions and assumes
+ * all coordinates are given with respect to the origin of the coordinate
+ * system.
```
I don't understand what this means. What else would I give the coordinates relative to?

knelli2 replied:
This was to try and disambiguate the origin from the center of the skew map, but this does make it confusing. I can remove it when I squash.

nilsdeppe replied:
Ah, okay. That's what I assumed but was only able to figure it out because I reviewed Mark's Spec PR :) Thanks for removing it!

nilsvu commented on src/Domain/CoordinateMaps/TimeDependent/Skew.cpp:
```diff
@@ -0,0 +1,424 @@
+std::optional<std::array<double, 3>> Skew::inverse(
+    const std::array<double, 3>& target_coords, const double time,
+    const std::unordered_map<
+        std::string, std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>>&
```
(optional) Can use `domain::FunctionsOfTimeMap` here as well

nilsvu reviewed (dismissed):
LGTM please squash
